// webdriverio/build/utils/interception/devtools.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _fs = _interopRequireDefault(require("fs"));

var _path = _interopRequireDefault(require("path"));

var _atob = _interopRequireDefault(require("atob"));

var _minimatch = _interopRequireDefault(require("minimatch"));

var _utils = require("@wdio/utils");

var _ = _interopRequireDefault(require("./"));

var _2 = require("..");

var _constants = require("../../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

class DevtoolsInterception extends _.default {
  static handleRequestInterception(client, mocks) {
    return async event => {
      const responseHeaders = event.responseHeaders.reduce((headers, {
        name,
        value
      }) => {
        headers[name] = value;
        return headers;
      }, {});
      const {
        requestId,
        request,
        responseStatusCode
      } = event;

      for (const mock of mocks) {
        if (!(0, _minimatch.default)(request.url, mock.url)) {
          continue;
        }

        request.statusCode = responseStatusCode;
        request.responseHeaders = _objectSpread({}, responseHeaders);

        if (filterMethod(request.method, mock.filterOptions.method) || filterHeaders(request.headers, mock.filterOptions.requestHeaders) || filterHeaders(responseHeaders, mock.filterOptions.headers) || filterRequest(request.postData, mock.filterOptions.postData) || filterStatusCode(responseStatusCode, mock.filterOptions.statusCode)) {
          continue;
        }

        const {
          body,
          base64Encoded
        } = await client.send('Fetch.getResponseBody', {
          requestId
        }).catch(() => ({}));
        request.body = base64Encoded ? (0, _atob.default)(body) : body;
        request.body = responseHeaders['Content-Type'] && responseHeaders['Content-Type'].includes('application/json') ? JSON.parse(request.body) : request.body;
        mock.matches.push(request);

        if (mock.respondOverwrites.length === 0) {
          continue;
        }

        const {
          errorReason,
          overwrite,
          params
        } = mock.respondOverwrites[0].sticky ? mock.respondOverwrites[0] : mock.respondOverwrites.shift();

        if (overwrite !== undefined) {
          let newBody = overwrite;

          if (typeof overwrite === 'function') {
            newBody = await overwrite(request, client);
          }

          if (typeof newBody !== 'string') {
            newBody = JSON.stringify(newBody);
          }

          let responseCode = params.statusCode || responseStatusCode;
          let responseHeaders = [...event.responseHeaders, ...Object.entries(params.headers || {}).map(([key, value]) => {
            key, value;
          })];
          const responseFilePath = _path.default.isAbsolute(newBody) ? newBody : _path.default.join(process.cwd(), newBody);

          if (_fs.default.existsSync(responseFilePath) && (0, _utils.canAccess)(responseFilePath)) {
            newBody = _fs.default.readFileSync(responseFilePath).toString();
          } else if (newBody.startsWith('http')) {
            responseCode = 301;
            responseHeaders = responseHeaders.filter(({
              name
            }) => name.toLowerCase() !== 'location');
            responseHeaders.push({
              name: 'Location',
              value: newBody
            });
          }

          request.mockedResponse = newBody;
          return client.send('Fetch.fulfillRequest', {
            requestId,
            responseCode,
            responseHeaders,
            body: Buffer.from(newBody).toString('base64')
          });
        }

        if (errorReason) {
          return client.send('Fetch.failRequest', {
            requestId,
            errorReason
          });
        }
      }

      return client.send('Fetch.continueRequest', {
        requestId
      });
    };
  }

  get calls() {
    return this.matches;
  }

  clear() {
    this.matches = [];
  }

  restore() {
    this.clear();
    this.respondOverwrites = [];
  }

  respond(overwrite, params = {}) {
    this.respondOverwrites.push({
      overwrite,
      params,
      sticky: true
    });
  }

  respondOnce(overwrite, params = {}) {
    this.respondOverwrites.push({
      overwrite,
      params
    });
  }

  abort(errorReason, sticky = true) {
    if (typeof errorReason !== 'string' || !_constants.ERROR_REASON.includes(errorReason)) {
      throw new Error(`Invalid value for errorReason, allowed are: ${_constants.ERROR_REASON.join(', ')}`);
    }

    this.respondOverwrites.push({
      errorReason,
      sticky
    });
  }

  abortOnce(errorReason) {
    this.abort(errorReason, false);
  }

}

exports.default = DevtoolsInterception;

const filterMethod = (method, expected) => {
  return expected && expected.toLowerCase() !== method.toLowerCase();
};

const filterHeaders = (responseHeaders, expected) => {
  return expected && !(0, _2.containsHeaderObject)(responseHeaders, expected);
};

const filterRequest = (postData, expected) => {
  if (typeof expected === 'undefined') {
    return false;
  }

  if (typeof expected === 'function') {
    return expected(postData) !== true;
  }

  return postData !== expected;
};

const filterStatusCode = (statusCode, expected) => {
  return typeof expected === 'number' && statusCode !== expected;
};
